iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -0,0 +1,225 @@
+
+                                InitializeClient();
+                                break;
+
+                            case ConnectionStatusChangeReason.Communication_Error:
+                                _logger.LogWarning("### The DeviceClient has been disconnected due to a non-retry-able exception. Inspect the exception for details." +
+                                    "\nIf you want to perform more operations on the device client, you should dispose (DisposeAsync()) and then open (OpenAsync()) the client.");
+
+                                InitializeClient();
+                                break;
+
+                            default:
+                                _logger.LogError("### This combination of ConnectionStatus and ConnectionStatusChangeReason is not expected, contact the client library team with logs.");
+                                    break;
+
+                        }
+                    }
+
+                    break;
+
+                default:
+                    _logger.LogError("### This combination of ConnectionStatus and ConnectionStatusChangeReason is not expected, contact the client library team with logs.");
+                    break;
+            }
+        }
+
+        public async Task RunSampleAsync()
+        {
+            await Task.WhenAll(SendMessagesAsync(), ReceiveMessagesAsync());
+        }
+
+        private void InitializeClient()
+        {
+            // If the client reports Connected status, it is already in operational state.
+            if (s_connectionStatus != ConnectionStatus.Connected)
+            {
+                lock (_lock)
+                {
+                    _logger.LogDebug($"Attempting to initialize the client instance, current status={s_connectionStatus}");
+
+                    // If the device client instance has been previously initialized, then dispose it.
+                    // The s_wasEverConnected variable is required to store if the client ever reported Connected status.
+                    if (s_wasEverConnected && s_connectionStatus == ConnectionStatus.Disconnected)
+                    {
+                        s_deviceClient?.Dispose();
+                        s_wasEverConnected = false;
+                    }
+
+                    s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionString, _transportType);
+                    s_deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangeHandler);
+                    s_deviceClient.OperationTimeoutInMilliseconds = (uint)s_operationTimeout.TotalMilliseconds;
+                    _logger.LogDebug($"Initialized the client instance.");
+                }
+            }
+        }
+
+        private async Task SendMessagesAsync()
+        {
+            int count = 0;
+            while (true)
```

Comment:
So the only exit condition is an unhandled exception?

Reply:
yes, since I wanted it to be long running (in order to be able to simulate exception and recovery), I did not put a limit on count/ time spent. I had a console input based approach, but while testing it, it was too cumbersome.

Reply:
How about catching control-c in the console, and signaling a CancellationToken?

Reply:
sure, I prefer this to setting a pre-determined timeout, I'll look into that

Reply:
https://github.com/Azure/azure-digital-twins/blob/private-preview/Demo/DeviceSimulator/DeviceSimulator/Program.cs